File: libs/prediction/src/postprocess.cpp
#include "pwb/prediction/postprocess.hpp"

#include <algorithm>
#include <filesystem>
#include <fstream>
#include <set>
#include <sstream>

#include "pwb/ingest/py_compat.hpp"
#include "pwb/ingest/well_parsers.hpp"
#include "pwb/interchange/unicode.hpp"
#include "pwb/prediction/errors.hpp"
#include "python_compat.hpp"

namespace pwb::prediction {

namespace {

using detail::is_word_char;
using detail::keep_word_chars;
using detail::py_finite_number;
using detail::py_int;
using detail::py_isclose;
using detail::py_or;
using detail::py_path_name;
using detail::py_path_stem;
using detail::py_repr;
using detail::py_round6;
using detail::py_str;
using detail::py_strip;
using detail::py_truthy;

constexpr double kEpsilon = 1e-6;
constexpr const char* kUnspecifiedStratum = "未标定层位";

const Json kNull;

const Json& get(const Json& obj, const char* key) {
    if (!obj.is_object()) return kNull;
    const auto it = obj.find(key);
    return it == obj.end() ? kNull : *it;
}

// _bounds(item): (top, bottom) when both finite and bottom - top > eps.
std::optional<std::pair<double, double>> bounds(const Json& item) {
    const auto top = py_finite_number(get(item, "top"));
    const auto bottom = py_finite_number(get(item, "bottom"));
    if (!top || !bottom || *bottom - *top <= kEpsilon) return std::nullopt;
    return std::pair{*top, *bottom};
}

// _normalize_boundaries(boundaries): dedup by (round6(depth), stripped
// name), keep first; then collapse to one record per depth, keeping the
// first in (depth, name) sort order.
Json normalize_boundaries(const Json& boundaries) {
    std::vector<Json> deduped;
    std::set<std::pair<double, std::string>> seen;
    if (boundaries.is_array()) {
        for (const auto& boundary : boundaries) {
            if (!boundary.is_object()) continue;
            const auto depth = py_finite_number(get(boundary, "depth"));
            if (!depth) continue;
            const std::string name = py_strip(
                py_str(py_or(get(boundary, "name"), Json(kUnspecifiedStratum))));
            const std::pair key{py_round6(*depth), name};
            if (seen.insert(key).second) {
                deduped.push_back(Json{
                    {"name", name.empty() ? kUnspecifiedStratum : name},
                    {"depth", py_round6(*depth)},
                    {"source", py_str(py_or(get(boundary, "source"), Json("")))},
                });
            }
        }
    }
    std::sort(deduped.begin(), deduped.end(), [](const Json& a, const Json& b) {
        const auto da = a.at("depth").get<double>();
        const auto db = b.at("depth").get<double>();
        if (da != db) return da < db;
        return a.at("name").get_ref<const std::string&>() <
               b.at("name").get_ref<const std::string&>();
    });
    Json out = Json::array();
    std::set<double> seen_depths;
    for (const auto& b : deduped) {
        if (seen_depths.insert(b.at("depth").get<double>()).second) out.push_back(b);
    }
    return out;
}

// _split_bounds: cut strictly inside (top+eps, bottom-eps); keep segments
// thicker than eps.
std::vector<std::pair<double, double>> split_bounds(
    double top, double bottom, const Json& boundaries) {
    std::vector<double> points{top};
    for (const auto& b : boundaries) {
        const double d = b.at("depth").get<double>();
        if (top + kEpsilon < d && d < bottom - kEpsilon) points.push_back(d);
    }
    points.push_back(bottom);
    std::vector<std::pair<double, double>> out;
    for (size_t i = 0; i + 1 < points.size(); ++i) {
        if (points[i + 1] - points[i] > kEpsilon) {
            out.emplace_back(points[i], points[i + 1]);
        }
    }
    return out;
}

// _stratum_at: index/name of the deepest boundary at or below `depth`.
std::pair<long long, std::string> stratum_at(double depth, const Json& boundaries) {
    long long index = 0;
    std::string name = kUnspecifiedStratum;
    for (const auto& b : boundaries) {
        if (b.at("depth").get<double>() > depth + kEpsilon) break;
        ++index;
        name = py_str(b.at("name"));
    }
    return {index, name};
}

// _display_probability: clamp then whole-percent half-even, stored as the
// decimal (53% -> 0.53). int(f"{v:.0%}"[:-1]) == nearbyint(v*100): both
// round the same double half-even.
double display_probability(const Json& value) {
    const auto p = py_finite_number(value);
    if (!p) return 0.0;
    const double clamped = std::max(0.0, std::min(1.0, *p));
    return std::nearbyint(clamped * 100.0) / 100.0;
}

// _touches: |left.bottom - right.top| <= eps on the stored (rounded) bounds.
bool touches(const Json& left, const Json& right) {
    const auto lb = bounds(left);
    const auto rb = bounds(right);
    return lb && rb && std::fabs(lb->second - rb->first) <= kEpsilon;
}

// _well_key: Path(str(v or "")).stem.casefold() minus non-\w characters.
std::string well_key(const Json& value) {
    const std::string name = py_path_stem(py_str(py_or(value, Json(""))));
    return keep_word_chars(pwb::interchange::casefold_utf8(name));
}

}  // namespace

PostprocessResult postprocess_prediction_regions(
    const Json& regions, const Json& formation_boundaries) {
    const Json normalized = normalize_boundaries(formation_boundaries);

    std::vector<Json> source;
    if (regions.is_array()) {
        for (const auto& region : regions) {
            if (region.is_object() && bounds(region)) source.push_back(region);
        }
    }
    std::stable_sort(source.begin(), source.end(), [](const Json& a, const Json& b) {
        return *bounds(a) < *bounds(b);
    });

    struct SplitItem {
        Json item;
        long long layer_index;
        std::string layer_name;
    };
    std::vector<SplitItem> split;
    for (const auto& region : source) {
        const auto [top, bottom] = *bounds(region);
        for (const auto& [seg_top, seg_bottom] :
             split_bounds(top, bottom, normalized)) {
            Json item = region;
            item["top"] = py_round6(seg_top);
            item["bottom"] = py_round6(seg_bottom);
            auto [layer_index, layer_name] =
                stratum_at((seg_top + seg_bottom) / 2.0, normalized);
            if (!normalized.empty()) item["stratigraphic_unit"] = layer_name;
            split.push_back({std::move(item), layer_index, std::move(layer_name)});
        }
    }

    std::vector<SplitItem> merged;
    for (auto& entry : split) {
        const double probability = display_probability(get(entry.item, "probability"));
        const std::string facies = py_strip(py_str(py_or(get(entry.item, "facies"), Json(""))));
        if (!merged.empty()) {
            auto& prev = merged.back();
            const std::string prev_facies =
                py_strip(py_str(py_or(get(prev.item, "facies"), Json(""))));
            if (!facies.empty() && prev_facies == facies &&
                display_probability(get(prev.item, "probability")) == probability &&
                prev.layer_index == entry.layer_index &&
                touches(prev.item, entry.item)) {
                prev.item["bottom"] = entry.item["bottom"];
                prev.item["probability"] = probability;
                const Json& prev_count = prev.item.contains("merged_sample_count")
                    ? prev.item["merged_sample_count"] : Json(1);
                const Json& item_count = entry.item.contains("merged_sample_count")
                    ? entry.item["merged_sample_count"] : Json(1);
                prev.item["merged_sample_count"] =
                    py_int(py_or(prev_count, Json(1))) + py_int(py_or(item_count, Json(1)));
                continue;
            }
        }
        Json copied = entry.item;
        copied["probability"] = probability;
        merged.push_back({std::move(copied), entry.layer_index, std::move(entry.layer_name)});
    }

    Json records = Json::array();
    long long index = 0;
    for (auto& entry : merged) {
        entry.item["region_id"] =
            "inference_api_post_" + std::to_string(++index);
        records.push_back(std::move(entry.item));
    }
    Json summary = {
        {"applied", true},
        {"confidence_display_precision", "1%"},
        {"raw_region_count", static_cast<long long>(source.size())},
        {"split_region_count", static_cast<long long>(split.size())},
        {"postprocessed_region_count", static_cast<long long>(records.size())},
        {"formation_boundary_count", static_cast<long long>(normalized.size())},
    };
    return {std::move(records), std::move(summary)};
}

ResolvedBoundaries resolve_formation_boundaries(
    const Json& well_name, const Json& well_log, const Json& inputs) {
    const std::string target = well_key(well_name);
    Json boundaries = Json::array();
    Json diagnostics = Json::array();

    if (!inputs.is_object()) {
        // (inputs or {}).values(): truthy non-mapping -> AttributeError.
        if (py_truthy(inputs)) {
            throw AttributeError("'" + detail::py_type_name(inputs) +
                                 "' object has no attribute 'values'");
        }
    } else {
        for (const auto& [key, info] : inputs.items()) {
            if (!info.is_object()) {
                throw AttributeError("'" + detail::py_type_name(info) +
                                     "' object has no attribute 'get'");
            }
            if (py_str(py_or(get(info, "asset_type"), Json(""))) !=
                "well_stratification") {
                continue;
            }
            const std::string path_str = py_str(py_or(get(info, "path"), Json("")));
            const std::filesystem::path path(path_str);
            const Json& name_v = get(info, "name");
            const std::string label = py_truthy(name_v)
                ? py_str(name_v) : py_path_name(path_str);
            if (!std::filesystem::is_regular_file(path)) {
                diagnostics.push_back("井分层文件不可读取: " + label);
                continue;
            }
            // Python parse_well_tops reads utf-8 errors="replace"; the C++
            // ingest parser is equally tolerant, so the Python
            // "井分层解析失败 {name}: {Class}" branch is unreachable here
            // (21-decisions.md D8).
            std::ifstream in(path, std::ios::binary);
            std::ostringstream buf;
            buf << in.rdbuf();
            const auto text = pwb::ingest::decode_utf8(buf.str(), false);
            const auto rows = pwb::ingest::parse_well_tops_text(text.value_or(""));
            for (const auto& row : rows) {
                if (well_key(Json(row.well_name)) == target) {
                    boundaries.push_back(Json{
                        {"name", row.top_name.empty()
                                     ? std::string(kUnspecifiedStratum)
                                     : row.top_name},
                        {"depth", row.md},
                        {"source", "well_stratification"},
                    });
                }
            }
        }
    }

    // well_log is a JSON attribute bag: {"intervals": {"formation": [...]}}.
    // Non-object values behave like getattr() misses -> empty.
    Json formations = Json::array();
    if (well_log.is_object()) {
        const Json& intervals = get(well_log, "intervals");
        if (intervals.is_object() && get(intervals, "formation").is_array()) {
            formations = get(intervals, "formation");
        }
    }
    for (const auto& interval : formations) {
        const auto depth = py_finite_number(get(interval, "top"));
        if (!depth) continue;
        const Json& name_v = get(interval, "name");
        boundaries.push_back(Json{
            {"name", py_truthy(name_v) ? py_str(name_v)
                                       : std::string(kUnspecifiedStratum)},
            {"depth", *depth},
            {"source", "las_formation"},
        });
    }

    return {normalize_boundaries(boundaries), std::move(diagnostics)};
}

}  // namespace pwb::prediction

File: legacy/python_reference/product/paleo_workbench/prediction/postprocess.py
"""Deterministic post-processing for persisted online facies predictions.

The inference API returns one categorical prediction at each sampled depth.
Those samples are first normalized into non-overlapping cells, then this
module reduces only *adjacent* compatible cells.  Formation tops are hard
boundaries: a reduction must never bridge a stratigraphic unit.
"""

from __future__ import annotations

import math
import re
from pathlib import Path
from typing import Any

from paleo_workbench.resources.well_tops_parser import parse_well_tops


_EPSILON = 1e-6
_UNSPECIFIED_STRATUM = "未标定层位"


def postprocess_prediction_regions(
    regions: list[dict[str, Any]] | None,
    *,
    formation_boundaries: list[dict[str, Any]] | None = None,
) -> tuple[list[dict[str, Any]], dict[str, Any]]:
    """Merge compatible adjacent prediction cells without crossing a top.

    Compatibility intentionally uses the same whole-percent precision shown by
    the confidence heatmap. Remote confidences are continuous floats, so
    requiring bitwise equality would almost never reduce a sampled result
    despite appearing identical to the operator. The persisted merged
    probability is the canonical displayed probability (for example, 53%
    becomes 0.53).

    Every supplied formation boundary splits a cell before the merge pass.
    This also protects against the uncommon case where a top falls inside a
    remote sample cell rather than exactly on its midpoint boundary.
    """
    normalized_boundaries = _normalize_boundaries(formation_boundaries)
    source = [
        dict(region)
        for region in (regions or [])
        if isinstance(region, dict) and _bounds(region) is not None
    ]
    source.sort(key=lambda item: (_bounds(item) or (math.inf, math.inf)))

    split: list[tuple[dict[str, Any], int, str]] = []
    for region in source:
        top, bottom = _bounds(region) or (0.0, 0.0)
        for segment_top, segment_bottom in _split_bounds(
            top, bottom, normalized_boundaries
        ):
            item = dict(region)
            item["top"] = round(segment_top, 6)
            item["bottom"] = round(segment_bottom, 6)
            layer_index, layer_name = _stratum_at(
                (segment_top + segment_bottom) / 2.0, normalized_boundaries
            )
            if normalized_boundaries:
                item["stratigraphic_unit"] = layer_name
            split.append((item, layer_index, layer_name))

    merged: list[tuple[dict[str, Any], int, str]] = []
    for item, layer_index, layer_name in split:
        probability = _display_probability(item.get("probability"))
        facies = str(item.get("facies") or "").strip()
        previous = merged[-1] if merged else None
        if (
            previous is not None
            and facies
            and str(previous[0].get("facies") or "").strip() == facies
            and _display_probability(previous[0].get("probability")) == probability
            and previous[1] == layer_index
            and _touches(previous[0], item)
        ):
            previous[0]["bottom"] = item["bottom"]
            previous[0]["probability"] = probability
            previous[0]["merged_sample_count"] = int(
                previous[0].get("merged_sample_count", 1) or 1
            ) + int(item.get("merged_sample_count", 1) or 1)
            continue

        copied = dict(item)
        copied["probability"] = probability
        merged.append((copied, layer_index, layer_name))

    records: list[dict[str, Any]] = []
    for index, (item, _layer_index, _layer_name) in enumerate(merged, start=1):
        item["region_id"] = f"inference_api_post_{index}"
        records.append(item)
    return records, {
        "applied": True,
        "confidence_display_precision": "1%",
        "raw_region_count": len(source),
        "split_region_count": len(split),
        "postprocessed_region_count": len(records),
        "formation_boundary_count": len(normalized_boundaries),
    }


def resolve_formation_boundaries(
    well_name: str,
    *,
    well_log: Any = None,
    inputs: dict[str, dict[str, Any]] | None = None,
) -> tuple[list[dict[str, Any]], list[str]]:
    """Resolve per-well formation tops from catalogued data and LAS metadata.

    Catalogued well_stratification inputs are the primary source because they
    describe project correlation tops. Embedded LAS formation intervals are
    added as a secondary source. A malformed optional stratification file
    never invalidates a successful remote inference; its diagnostic is saved
    with the post-processing metadata instead.
    """
    target = _well_key(well_name)
    boundaries: list[dict[str, Any]] = []
    diagnostics: list[str] = []

    for info in (inputs or {}).values():
        if str(info.get("asset_type") or "") != "well_stratification":
            continue
        path = Path(str(info.get("path") or ""))
        if not path.is_file():
            diagnostics.append(f"井分层文件不可读取: {info.get('name') or path.name}")
            continue
        try:
            rows = parse_well_tops(path)
        except Exception as exc:  # optional context must not discard a prediction
            diagnostics.append(
                f"井分层解析失败 {info.get('name') or path.name}: {exc.__class__.__name__}"
            )
            continue
        for row in rows:
            if _well_key(row.well_name) == target:
                boundaries.append(
                    {
                        "name": str(row.top_name or _UNSPECIFIED_STRATUM),
                        "depth": float(row.md),
                        "source": "well_stratification",
                    }
                )

    intervals = getattr(well_log, "intervals", None)
    formations = list(getattr(intervals, "formation", None) or [])
    for interval in formations:
        depth = _finite_number(getattr(interval, "top", None))
        if depth is not None:
            boundaries.append(
                {
                    "name": str(getattr(interval, "name", "") or _UNSPECIFIED_STRATUM),
                    "depth": depth,
                    "source": "las_formation",
                }
            )

    return _normalize_boundaries(boundaries), diagnostics


def _normalize_boundaries(
    boundaries: list[dict[str, Any]] | None,
) -> list[dict[str, Any]]:
    deduped: dict[tuple[float, str], dict[str, Any]] = {}
    for boundary in boundaries or []:
        if not isinstance(boundary, dict):
            continue
        depth = _finite_number(boundary.get("depth"))
        if depth is None:
            continue
        name = str(boundary.get("name") or _UNSPECIFIED_STRATUM).strip()
        key = (round(depth, 6), name)
        deduped.setdefault(
            key,
            {
                "name": name or _UNSPECIFIED_STRATUM,
                "depth": round(depth, 6),
                "source": str(boundary.get("source") or ""),
            },
        )
    # Multiple labels at exactly the same depth still mean one hard boundary;
    # preserve a deterministic label rather than creating zero-thickness units.
    by_depth: dict[float, dict[str, Any]] = {}
    for boundary in sorted(deduped.values(), key=lambda item: (item["depth"], item["name"])):
        by_depth.setdefault(float(boundary["depth"]), boundary)
    return list(by_depth.values())


def _split_bounds(
    top: float, bottom: float, boundaries: list[dict[str, Any]]
) -> list[tuple[float, float]]:
    cuts = [
        float(boundary["depth"])
        for boundary in boundaries
        if top + _EPSILON < float(boundary["depth"]) < bottom - _EPSILON
    ]
    points = [top, *cuts, bottom]
    return [
        (start, end)
        for start, end in zip(points, points[1:])
        if end - start > _EPSILON
    ]


def _stratum_at(
    depth: float, boundaries: list[dict[str, Any]]
) -> tuple[int, str]:
    index = 0
    name = _UNSPECIFIED_STRATUM
    for boundary in boundaries:
        if float(boundary["depth"]) > depth + _EPSILON:
            break
        index += 1
        name = str(boundary["name"])
    return index, name


def _display_probability(value: Any) -> float:
    probability = _finite_number(value)
    if probability is None:
        return 0.0
    probability = max(0.0, min(1.0, probability))
    # Keep the stored number in exact lockstep with the percentage rendered to
    # the operator. This reproduces Python percent-format rounding exactly.
    return int(f"{probability:.0%}"[:-1]) / 100.0


def _bounds(item: dict[str, Any]) -> tuple[float, float] | None:
    top = _finite_number(item.get("top"))
    bottom = _finite_number(item.get("bottom"))
    if top is None or bottom is None or bottom - top <= _EPSILON:
        return None
    return top, bottom


def _touches(left: dict[str, Any], right: dict[str, Any]) -> bool:
    left_bounds = _bounds(left)
    right_bounds = _bounds(right)
    return bool(
        left_bounds is not None
        and right_bounds is not None
        and abs(left_bounds[1] - right_bounds[0]) <= _EPSILON
    )


def _finite_number(value: Any) -> float | None:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if math.isfinite(number) else None


def _well_key(value: Any) -> str:
    name = Path(str(value or "")).stem.casefold()
    return re.sub(r"[^\w]", "", name, flags=re.UNICODE)


__all__ = [
    "postprocess_prediction_regions",
    "resolve_formation_boundaries",
]
